src/Cns.ts

```diff
@@ -1,35 +1,71 @@
-import { EthereumNamingService } from './EthereumNamingService';
-import { BlockhanNetworkUrlMap, ProxyReaderMap } from './types';
-import { default as proxyReaderAbi } from './cns/contract/proxyReader';
-import { default as resolverInterface } from './cns/contract/resolver';
+import { BlockhanNetworkUrlMap, CnsSupportedNetwork, ProxyReaderMap } from './types';
+import { default as proxyReaderAbi } from './contracts/cns/proxyReader';
+import { default as resolverInterface } from './contracts/cns/resolver';
 import ResolutionError, { ResolutionErrorCode } from './errors/resolutionError';
-import Contract from './utils/contract';
+import EthereumContract from './contracts/EthereumContract';
 import standardKeys from './utils/standardKeys';
-import { getStartingBlock, isLegacyResolver, isNullAddress } from './utils';
-import {
-  SourceDefinition,
-  NamingServiceName,
-  ResolutionResponse,
-  CryptoRecords,
-  DomainData,
-} from './publicTypes';
+import { constructRecords, isNullAddress } from './utils';
+import { CnsSource, CryptoRecords, DomainData, NamingServiceName, Provider } from './types/publicTypes';
 import { isValidTwitterSignature } from './utils/TwitterSignatureValidator';
 import NetworkConfig from './config/network-config.json';
+import FetchProvider from './FetchProvider';
+import { eip137Namehash } from './utils/namehash';
+import { NamingService } from './NamingService';
+import ConfigurationError, { ConfigurationErrorCode } from './errors/configurationError';
 
-export default class Cns extends EthereumNamingService {
-  static TwitterVerificationAddress = '0x12cfb13522F13a78b650a8bCbFCf50b7CB899d82';
-  static ProxyReaderMap: ProxyReaderMap = getProxyReaderMap();
+/**
+ * @internal
+ */
+export default class Cns extends NamingService {
+  static readonly ProxyReaderMap: ProxyReaderMap = getProxyReaderMap();
 
-  constructor(source: SourceDefinition = {}) {
-    super(source, NamingServiceName.CNS);
+  static readonly UrlMap: BlockhanNetworkUrlMap = {
+    1: 'https://mainnet.infura.io/v3/c4bb906ed6904c42b19c95825fe55f39',
+    4: 'https://rinkeby.infura.io/v3/c4bb906ed6904c42b19c95825fe55f39',
+  };
+  static readonly NetworkNameMap = {
+    mainnet: 1,
+    rinkeby: 4
+  };
+
+  readonly name: NamingServiceName = NamingServiceName.CNS;
+  readonly network: number;
+  readonly url: string | undefined;
+  readonly provider: Provider;
+  readonly readerContract: EthereumContract;
+
+  constructor(source?: CnsSource) {
+    super();
+    if (!source) {
+      source = {
+        url: Cns.UrlMap[1],
+        network: "mainnet",
+      };
+    }
+    if (!source.network || !CnsSupportedNetwork.guard(source.network)) {
+      throw new ConfigurationError(ConfigurationErrorCode.UnspecifiedNetwork, {
+        method: NamingServiceName.CNS,
+      });
```

Comment:
Passing an unsupported network results in `UnspecifiedNetwork` error.

I also feel like there's a lack of tests for our name service constructors.